### PR TITLE
Dapt data curation tutorial fuzzy and semantic dedupe

### PR DESCRIPTION
**Pull request by ruchaa-apte**
## Description
The PR adds 
- Versions for installations edits in requirements.txt for pip packages
- `fuzzy_dedupe` module implementation to DAPT data curator tutorial for chipNeMo. It performs Fuzzy Deduplication on code as well as text documents from Wikipedia and Arxiv sources.
- `semantic_dedupe` module implementation to DAPT data curator tutorial for chipNeMo. It performs Semantic Deduplication on text documents from Wikipedia and Arxiv sources.
## Usage
```
cd NeMo-Curator/tutorials/dapt-curation/code/
python main.py
```
## Checklist
- [ X] I am familiar with the [Contributing Guide](https://github.com/NVIDIA/NeMo-Curator/blob/main/CONTRIBUTING.md).
- [ X] New or Existing tests cover these changes.
- [ X] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by ruchaa-apte** [comment]
pre-commit.ci autofix

**Comment by ayushdg** [comment]
Looks like the commits are not gpg signed. You might need to recommit your changes with commit signing enabled `-Ss` flags during git commit. `-S` for signing and `-s` for sigoff.
More info on signing commits here: https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits

**Comment by ruchaa-apte** [comment]
pre-commit.ci autofix

**Comment by ruchaa-apte** [comment]
> Looks like the commits are not gpg signed. You might need to recommit your changes with commit signing enabled `-Ss` flags during git commit. `-S` for signing and `-s` for sigoff. More info on signing commits here: https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits

Hi Ayush, thank you for pointing that out. Looks like me email address linked to signing off was not correct. I modified it and signed off the commit.

**Comment by ruchaa-apte** [comment]
pre-commit.ci autofix

File: tutorials/dapt-curation/README.md [hunk 1]
```diff
@@ -45,8 +45,10 @@ The tutorial follows the steps below:<br>
 
 After installing the NeMo Curator package, install the dependencies and run:
```
Comment (Maghoumi) [inline tutorials/dapt-curation/README.md:47]: In the steps above, please also clarify that there are now different versions of deduplication (it's currently only mentioning `exact`)

File: tutorials/dapt-curation/code/main.py [hunk 2]
```diff
@@ -234,11 +268,13 @@ def main():
     args = ArgumentHelper(parser).add_distributed_args().parse_args()
     # Limit the total number of workers to ensure we don't run out of memory.
     args.n_workers = min(args.n_workers, 8)
+    args.device = "gpu"
```
Comment (Maghoumi) [inline tutorials/dapt-curation/code/main.py:271]: Instead of hardcoding this, we need to let the users decide what device to use. Then under the hood, ensure that GPU processing steps only run if the user has passed `--device gpu`

File: tutorials/dapt-curation/code/utils.py [hunk 3]
```diff
@@ -282,6 +296,79 @@ def dedupe(dataset: DocumentDataset) -> DocumentDataset:
...
+    Removes near-duplicate documents and code lines
+
+    Args:
+        dataset (DocumentDataset): The dataset containing documents.
+        type (str): Document type to process.
+
+    Returns:
+        DocumentDataset: The deduplicated dataset.
+    """
+    cache_dir = f"./workspace/fuzzy_dedupe_cache/{type}"
```
Comment (Maghoumi) [inline tutorials/dapt-curation/code/utils.py:310]: Instead of hardcoding, can you pass it into the function instead? Meaning

```
def fuzzy_dedupe(dataset: DocumentDataset, type: str = "text", cache_dir: str) -> DocumentDataset:
```

Then from inside the main curation loop, we can pass in a value that is relative to the script path, like:

```
CACHE_DIR = os.path.join(SCRIPT_DIR_PATH, "cache")
fuzzy_dedupe(..., cache=CACHE_DIR)

```

File: tutorials/dapt-curation/code/utils.py [hunk 4]
```diff
@@ -282,6 +296,79 @@ def dedupe(dataset: DocumentDataset) -> DocumentDataset:
...
+    Perform semantic deduplication on the given dataset.
+
+    Args:
+        dataset (DocumentDataset): The dataset containing documents.
+        type (str): Document type to process.
+
+    Returns:
+        The deduplicated DocumentDataset.
+    """
+    cache_dir = f"./workspace/semantic_dedupe/{type}"
```
Comment (Maghoumi) [inline tutorials/dapt-curation/code/utils.py:355]: Same as above.

**Comment by ruchaa-apte** [comment]
pre-commit.ci autofix

File: tutorials/dapt-curation/code/requirements.txt [hunk 5]
```diff
@@ -1,6 +1,7 @@
-arxiv
+arxiv==2.1.0
```
Comment (sarahyurick) [inline tutorials/dapt-curation/code/requirements.txt:1]: Why did you add the version pin here?

**Review by sarahyurick — APPROVED** [review]
LGTM!

File: tutorials/dapt-curation/code/requirements.txt (hunk 5 above)
Comment (ruchaa-apte) [reply tutorials/dapt-curation/code/requirements.txt:1]: This was the error I got previously which was resolved by the above fix
```
  File "/home/ldu/Projects/supervised_fine_tuning/code/docbuilder.py", line 420, in download
    search_result = arxiv.Client().results(arxiv.Search(id_list=[article_id]))
AttributeError: 'Client' object has no attribute 'results'
```

File: tutorials/dapt-curation/code/configs/text_semantic_dedupe_config.yaml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+num_files: 16
+id_col_name: "id"
+id_col_type: "str"
+input_column: "text"
+
+# Embeddings configuration
+embeddings_save_loc: "embeddings"
+embedding_model_name_or_path: "sentence-transformers/all-MiniLM-L6-v2"
+embedding_batch_size: 128
+embedding_max_mem_gb: 20
```
Comment (sarahyurick) [inline tutorials/dapt-curation/code/configs/text_semantic_dedupe_config.yaml:12]: https://github.com/NVIDIA/NeMo-Curator/pull/327 removed `id_col_name`, `id_col_type`, `input_column`, and `embedding_max_mem_gb` from the configuration file. It also renamed `id_col_name` to `id_column` and `id_col_type` to `id_column_type`. Please update. Thanks!

File: tutorials/dapt-curation/code/utils.py [hunk 7]
```diff
@@ -282,6 +296,71 @@ def dedupe(dataset: DocumentDataset) -> DocumentDataset:
...
+    """
+    partition_lengths = dataset.df.map_partitions(len).compute()
+    non_empty_partitions = [
+        i for i, length in enumerate(partition_lengths) if length > 0
+    ]
+    dataset.df = dataset.df.partitions[non_empty_partitions]
+
+    semdedup_config = SemDedupConfig.from_yaml(sem_dedupe_config_yaml_path)
+    expand_outdir_and_mkdir(semdedup_config.cache_dir)
+    semdup = SemDedup(semdedup_config)
```
Comment (sarahyurick) [inline tutorials/dapt-curation/code/utils.py:359]: Please confirm whether these default values are correct for your use case:

- `input_column="text"`
- `id_column="id"`
- `id_column_type="int"`

**Review by sarahyurick — APPROVED** [review]

**Comment by sarahyurick** [comment]
Hi @ruchaa-apte please sign all commits so that I can merge.

**Comment by ruchaa-apte** [comment]
pre-commit.ci autofix

**Comment by ruchaa-apte** [comment]
pre-commit.ci autofix

**Comment by sarahyurick** [comment]
Re-opened as https://github.com/NVIDIA/NeMo-Curator/pull/428.
